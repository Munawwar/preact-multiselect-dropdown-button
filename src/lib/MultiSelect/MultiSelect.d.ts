import { Component } from 'preact';

export type ListItem = {
  label: string,
  id: string,
  name?: string,
  checked: boolean
};

export type Props = {
  /**
   * List of items to be shown in the dropdown
   */
  list: ListItem[],
  dropdownButtonText: string,
  isRightAligned?: boolean,
  selectAllButtonText?: string,
  /**
   * Triggered immediately when items are checked (before clicking the 'apply' button)
   */
  onOptionChanged?: (list: ListItem[]) => void,
  /**
   * Triggered after selections are applied (by clicking the 'apply' button)
   */
  onSelectionApplied?: (list: ListItem[]) => void,
  resetButtonText?: string,
  applyButtonText?: string,
}

export type State = {
  isDropdownOpened: boolean,
  checkedItems: ListItem[],
};

export default class MultiSelect extends Component<Props, State> {}

