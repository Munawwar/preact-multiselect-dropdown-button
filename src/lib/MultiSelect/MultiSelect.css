
:root {
  /* MultiSelect.jsx */
  --MultiSelect-buttonHeight: 4.125rem; /* 66px if base font-size is 16px */
  --MultiSelect-selectAllButtonHeight: 3.125rem; /* 50px if base font-size is 16px */
  --MultiSelect-paddingList: 2.625rem; /* 42px if base font-size is 16px */
  --MultiSelect-footerHeight: 3.75rem; /* 60px if base font-size is 16px */

  /* MultiSelectBadge.jsx */
  --MultiSelect-badgeColor: #fff;
  --MultiSelect-backgroundColor: #000;
}

/* MultiSelect.jsx */

.MultiSelect-wrapper {
  position: relative;
  display: inline-block;
}

.MultiSelect-sectionWrapper {
  position: absolute;
  right: auto;
  min-width: 140px;
  padding: 1.25rem; /* 20px if base font-size is 16px */
  border: 1px #000 solid;
}
.MultiSelect-sectionWrapper--rightAligned {
  right: 0;
}

.MultiSelect-list {
  max-height: calc(
    100vh
    - var(--MultiSelect-buttonHeight)
    - var(--MultiSelect-paddingList)
    - var(--MultiSelect-selectAllButtonHeight)
  );
  margin: 0;
  overflow: auto;
  padding-left: 0;
  list-style: none;
}

.MultiSelect-list--hasFooter {
  max-height: calc(
    100vh
    - var(--MultiSelect-buttonHeight)
    - var(--MultiSelect-paddingList)
    - var(--MultiSelect-selectAllButtonHeight)
    - var(--MultiSelect-footerHeight)
  );
}


/* MultiSelectBadge.jsx */

.MultiSelect-badge {
  padding: 0.1875rem 0.5rem; /* 3px 8px if base font-size is 16px */
  margin-left: 0.3125rem;
  font-size: 0.85rem;
  font-weight: normal;
  line-height: 1;
  color: var(--MultiSelect-badgeColor);
  vertical-align: middle;
  background-color: var(--MultiSelect-backgroundColor);
  border-radius: 50%;
}


/* MultiSelectDropdown.jsx */

.MultiSelect-dropdownButton {
  height: 100%;
  padding: 1.25rem; /* 20px if base font-size is 16px */
  font-family: inherit;
  font-size: 1rem;
  cursor: pointer;
  background-color: transparent;
  border: 1px #000 solid;
  border-bottom: 0;
  box-shadow: 0 1px #000;
}
.MultiSelect-dropdownButton:after {
  content: '';
  display: inline-block;
  margin-left: 10px;
  padding: 3px;
  border: solid black;
  border-width: 0 3px 3px 0;
  transform: rotate(45deg);
  -webkit-transform: rotate(45deg);
}
.MultiSelect-dropdownButton--opened:after {
  transform: rotate(-135deg);
  -webkit-transform: rotate(-135deg);
}

/* MultiSelectFooter.jsx */

.MultiSelect-footer {
  display: flex;
  flex-direction: row-reverse;
  padding-top: 1.25rem; /* 20px if base font-size is 16px */
  border-top: 1px #ccc solid;
}

.MultiSelect-footerApplyButton {
  display: block;
  height: 40px;
  padding: 0 0.625rem;
  font-family: inherit;
  line-height: 40px;
  text-align: center;
  color: #fff;
  background-color: #000;
  border: none;
  border-radius: 0;
  cursor: pointer;
}


/* MultiSelectListButtons.jsx */

.MultiSelect-listButtonsWrapper {
  display: flex;
  justify-content: space-between;
}

.MultiSelect-listButtonsSelectAll {
  display: block;
  height: 40px;
  padding: 0 0.625rem;
  font-family: inherit;
  line-height: 40px;
  text-align: center;
  color: #fff;
  background-color: #000;
  border: none;
  border-radius: 0;
  cursor: pointer;
  height: 1.3em;
  line-height: 1em;
  color: #5e9ed6;
  background-color: transparent;
  margin-bottom: 10px;
  padding-left: 0;
}

.MultiSelect-listButtonsReset {
  display: block;
  height: 40px;
  padding: 0 0.625rem;
  font-family: inherit;
  line-height: 40px;
  text-align: center;
  color: #fff;
  background-color: #000;
  border: none;
  border-radius: 0;
  cursor: pointer;
  height: 1.3em;
  line-height: 1em;
  color: #5e9ed6;
  background-color: transparent;
  padding-right: 0;
}


/* MultiSelectListItem.jsx */

.MultiSelect-listItem:hover {
  background-color: #f0f0f0;
}

.MultiSelect-listItemCheckbox {
  visibility: hidden;
}

.MultiSelect-lisItemLabel {
  display: flex;
  justify-content: space-between;
  padding: 0.625rem 0; /* 10px if base font-size is 16px */
  margin: 1px;
  line-height: 1;
  white-space: nowrap;
  cursor: pointer;
}
.MultiSelect-lisItemLabel:focus {
  outline: 1px solid #5e9ed6;
}
.MultiSelect-lisItemLabel--checked:after {
  content: '';
  display: block;
  width: 5px;
  height: 10px;
  margin-right: 3px;
  border: solid #000;
  border-width: 0 3px 3px 0;
  transform: rotate(45deg);
}