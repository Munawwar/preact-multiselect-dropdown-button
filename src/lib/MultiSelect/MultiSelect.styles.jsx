import styled, { css } from 'styled-components';
import ListboxKeyNav from '../react-listbox-keyboard-navigation/ListboxKeyNav';

export const MultiSelectWrapper = styled('div')`
  position: relative;
  display: inline-block;
`;

export const MultiSelectListWrapper = styled('div')`
  position: absolute;
  right: ${props => (props.isRightAligned ? '0' : 'auto')};
  min-width: 140px;
  padding: 1.25rem; /* 20px if base font-size is 16px */
  border: 1px #000 solid;
`;

const buttonHeight = '4.125rem'; /* 66px if base font-size is 16px */
const selectAllButtonHeight = '3.125rem'; /* 50px if base font-size is 16px */
const paddingList = '2.625rem'; /* 42px if base font-size is 16px */
const footerHeight = '3.75rem'; /* 60px if base font-size is 16px */

export const MultiSelectList = styled(ListboxKeyNav)`
  max-height: ${props =>
    `calc(100vh - ${buttonHeight} - ${paddingList} - ${selectAllButtonHeight} - ${
      props.hasFooter ? footerHeight : '0px'
    })`};
  margin: 0;
  overflow: auto;
  padding-left: 0;
  list-style: none;
`;

export const buttonCSS = css`
  display: block;
  height: 40px;
  padding: 0 0.625rem;
  font-family: inherit;
  line-height: 40px;
  text-align: center;
  color: #fff;
  background-color: #000;
  border: none;
  border-radius: 0;
  cursor: pointer;
`;

export const MultiSelectAllButton = styled('button')`
  ${buttonCSS}
  margin-bottom: 10px;
`;

export const MultiSelectResetButton = styled('button')`
  ${buttonCSS}
`;

export const iconCSS = css`
  width: 0.75rem; /* 12px if base font-size is 16px */
  margin-right: 0.625rem; /* 10px if base font-size is 16px */
  vertical-align: baseline;
`;

export const MultiSelectListButtonsWrapper = styled('div')`
  display: flex;
  justify-content: space-between;
`;
