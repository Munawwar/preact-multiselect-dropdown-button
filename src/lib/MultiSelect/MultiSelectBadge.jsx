/**
 * @param {object} props
 * @param {string} props.text
 * @param {number} props.quantity
 */
const MultiSelectBadge = ({ text, quantity }) => (
  <>
    {text} <span className="PreactMultiSelect-badge">{quantity}</span>
  </>
);

export default MultiSelectBadge;
