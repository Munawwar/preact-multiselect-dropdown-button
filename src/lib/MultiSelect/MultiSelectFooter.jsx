/**
 * @param {object} props
 * @param {string} [props.applyButtonText = 'Apply']
 * @param {() => void} [props.handleApplyClick]
 */
const MultiSelectFooter = ({ applyButtonText = 'Apply', handleApplyClick }) => (
  <div className="MultiSelect-footer">
    <button className="MultiSelect-footerApplyButton" onClick={handleApplyClick}>
      {applyButtonText}
    </button>
  </div>
);

export default MultiSelectFooter;
