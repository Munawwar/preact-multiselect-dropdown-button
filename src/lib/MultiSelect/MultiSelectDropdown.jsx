import MultiSelectBadge from './MultiSelectBadge';

/**
 * @param {object} props
 * @param {string} props.text
 * @param {() => void} props.toggleDropdown
 * @param {number} props.quantity
 * @param {boolean} props.isOpened
 */
const MultiSelectDropdown = ({ text, toggleDropdown, quantity, isOpened }) => (
  <button
    className={`MultiSelect-dropdownButton${isOpened ? ' MultiSelect-dropdownButton--opened' : ''}`}
    onClick={toggleDropdown}
    aria-haspopup="true"
  >
    <MultiSelectBadge className="MultiSelect-badge" text={text} quantity={quantity} />
  </button>
);

export default MultiSelectDropdown;
