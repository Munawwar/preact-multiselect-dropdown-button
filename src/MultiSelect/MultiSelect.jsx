import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import ListboxKeyEvents from '../react-listbox-key-events/ListboxKeyEvents';
import MultiSelectDropdown from './MultiSelectDropdown';
import MultiSelectListItem from './MultiSelectListItem';
import MultiSelectFooter from './MultiSelectFooter';

const MultiSelectWrapper = styled('div')`
  position: relative;
`;

const MultiSelectListWrapper = styled('div')`
  position: absolute;
  right: ${props => props.isRightAligned ? '0' : 'auto'};
  min-width: 170px;
  padding: 1.25rem; /* 20px if base font-size is 16px */
  border: 1px #000 solid;
`;

const MultiSelectList = styled(ListboxKeyEvents)`
  padding-left: 0;
  list-style: none;
`;

class MultiSelect extends React.PureComponent {
  state = {
    isDropdownOpened: false,
    checkedItems: {}
  };

  constructor(props) {
    super(props);
    this.listItems = [];

    this.keyEvents = {
      up: () => {
        const activeElementIndex = this.listItems.findIndex(item => {
          return item === document.activeElement;
        });
        this.listItems[activeElementIndex - 1] && this.listItems[activeElementIndex - 1].focus();
      },
      down: () => {
        const activeElementIndex = this.listItems.findIndex(item => {
          return item === document.activeElement;
        });
        this.listItems[activeElementIndex + 1] && this.listItems[activeElementIndex + 1].focus();
      }
    };
  }

  componentDidMount() {
    const { list } = this.props;

    this.setState({
      checkedItems: list.reduce((acc, listItem) => {
        acc[listItem.name] = listItem.checked;
        return acc;
      }, {})
    });
  }

  toggleDropdown = () => {
    const { isDropdownOpened } = this.state;

    this.setState({ isDropdownOpened: !isDropdownOpened }, () => {
      if (this.state.isDropdownOpened) {
        this.listItems[0].focus();
      }
    });
  };

  handleInputChange = event => {
    const tag = event.target.tagName;
    const target = tag === 'LABEL' ? event.target.children[0] : event.target;
    // If user used the keyboard to select the label, we need to programatically check the checkbox child.
    // Also needed to work with screen readers
    if (tag === 'LABEL') {
      target.checked = !target.checked;
    }
    const { name, checked } = target;

    this.setState(prevState => ({
      checkedItems: { ...prevState.checkedItems, [name]: checked }
    }));
  };

  resetSelections = () => {
    this.setState(prevState => ({
      checkedItems: Object.keys(prevState.checkedItems).reduce((acc, listItemName) => {
        acc[listItemName] = false;
        return acc;
      }, {})
    }));
  };

  handleApplyClick = () => {
    const { onSelectionApplied } = this.props;
    const { checkedItems } = this.state;

    onSelectionApplied(checkedItems);
  };

  render() {
    const { isDropdownOpened, checkedItems } = this.state;
    const { list, dropdownButtonText, isRightAligned, onSelectionApplied, resetButtonText, applyButtonText } = this.props;
    const { handleInputChange, handleApplyClick } = this;
    const checkedItemsQuantity = Object.keys(checkedItems).filter(itemName => checkedItems[itemName]).length;

    return (
      <MultiSelectWrapper>
        <MultiSelectDropdown
          className="multiselect-section-wrapper"
          text={dropdownButtonText}
          quantity={checkedItemsQuantity}
          toggleDropdown={this.toggleDropdown}
          isOpened={isDropdownOpened}
        />
        {isDropdownOpened && (
          <MultiSelectListWrapper {...(isRightAligned ? {isRightAligned} : {})} className="multiselect-section-wrapper">
            <MultiSelectList
              role="listbox"
              className="multiselect-list"
              keyEvents={{...this.keyEvents}}
            >
              {list.map((listItem, index) => {
                const { label, id, name } = listItem;
                const checked = checkedItems[name];
                const key = `${id}-${index}`;

                return (
                  <MultiSelectListItem
                    className="multiselect-list-item"
                    label={label}
                    key={key}
                    id={id}
                    name={name}
                    handleInputChange={handleInputChange}
                    checked={checked}
                    ref={el => {
                      this.listItems[index] = el;
                      return el;
                    }}
                  />
                );
              })}
            </MultiSelectList>
            {onSelectionApplied && <MultiSelectFooter
              resetSelections={this.resetSelections}
              resetButtonText={resetButtonText}
              applyButtonText={applyButtonText}
              handleApplyClick={handleApplyClick}
            />}
          </MultiSelectListWrapper>
        )}
      </MultiSelectWrapper>
    );
  }
}

MultiSelect.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      checked: PropTypes.bool.isRequired
    })
  ).isRequired,
  dropdownButtonText: PropTypes.string.isRequired,
  isRightAligned: PropTypes.bool,
  onSelectionApplied: PropTypes.func,
  resetButtonText: PropTypes.string,
  applyButtonText: PropTypes.string
};

export default MultiSelect;
