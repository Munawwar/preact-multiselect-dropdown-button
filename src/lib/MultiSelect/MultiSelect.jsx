import { Component } from 'preact';
import ListboxKeyNav from '../react-listbox-keyboard-navigation/ListboxKeyNav';
import MultiSelectDropdown from './MultiSelectDropdown';
import MultiSelectListButtons from './MultiSelectListButtons';
import MultiSelectListItem from './MultiSelectListItem';
import MultiSelectFooter from './MultiSelectFooter';
import getMultiSelectHandlers, { getInitialCheckedItems } from './multiSelectHandler';

class MultiSelect extends Component {
  constructor(props) {
    super(props);
    this.setIsDropdownOpened = this.setIsDropdownOpened.bind(this);
    this.setCheckedItems = this.setCheckedItems.bind(this);

    this.listItems = [];

    this.state = {
      isDropdownOpened: false,
      checkedItems: getInitialCheckedItems(this.props.list || []),
    };
  }

  setIsDropdownOpened(value) {
    this.setState({ isDropdownOpened: value });
  }
  setCheckedItems(value) {
    this.setState({ checkedItems: value });
  }

  onUpdate(prevProps, prevState) {
    const { onOptionChanged } = this.props;
    const { isDropdownOpened, checkedItems } = this.state;

    if (isDropdownOpened && this.listItems[0] && (!prevState || isDropdownOpened !== prevState.isDropdownOpened)) {
      this.listItems[0].focus();
    }

    if (onOptionChanged && (!prevState || checkedItems !== prevState.checkedItems)) {
      onOptionChanged(checkedItems);
    }
  }

  componentDidMount() {
    this.onUpdate();
  }

  componentDidUpdate(prevProps, prevState) {
    this.onUpdate(prevProps, prevState);
  }

  render() {
    const {
      list,
      dropdownButtonText,
      isRightAligned,
      selectAllButtonText,
      onSelectionApplied,
      resetButtonText,
      applyButtonText,
    } = this.props;

    const { isDropdownOpened, checkedItems } = this.state;

    const { toggleDropdown, selectAll, resetSelections, handleInputChange } = getMultiSelectHandlers({
      isDropdownOpened,
      setIsDropdownOpened: this.setIsDropdownOpened,
      checkedItems,
      setCheckedItems: this.setCheckedItems,
    });

    this.listItems = [];
    const keyEvents = {
      up: () => {
        const activeElementIndex = this.listItems.findIndex((item) => item === document.activeElement);
        this.listItems[activeElementIndex - 1] && this.listItems[activeElementIndex - 1].focus();
      },
      down: () => {
        const activeElementIndex = this.listItems.findIndex((item) => item === document.activeElement);
        this.listItems[activeElementIndex + 1] && this.listItems[activeElementIndex + 1].focus();
      },
      home: () => {
        this.listItems[0] && this.listItems[0].focus();
      },
      end: () => {
        this.listItems[this.listItems.length - 1] && this.listItems[this.listItems.length - 1].focus();
      },
    };
    const checkedItemsQuantity = Object.keys(checkedItems).filter((itemName) => checkedItems[itemName]).length;

    const handleApplyClick = () => {
      onSelectionApplied(checkedItems);
    };

    return (
      <div className="MultiSelect-wrapper">
        <MultiSelectDropdown
          className="multiselect-button-dropdown"
          text={dropdownButtonText}
          quantity={checkedItemsQuantity}
          toggleDropdown={toggleDropdown}
          isOpened={isDropdownOpened}
        />
        {isDropdownOpened && (
          <div
            className={`MultiSelect-sectionWrapper${isRightAligned ? ' MultiSelect-sectionWrapper--rightAligned' : ''}`}
          >
            <MultiSelectListButtons
              selectAll={selectAll}
              selectAllButtonText={selectAllButtonText}
              resetSelections={resetSelections}
              resetButtonText={resetButtonText}
            />
            <ListboxKeyNav
              role="listbox"
              tag="ul"
              className={`MultiSelect-list${onSelectionApplied ? ' MultiSelect-list--hasFooter' : ''}`}
              keyEvents={{ ...keyEvents }}
            >
              {list.map((listItem, index) => {
                const { label, id, name } = listItem;
                const checked = checkedItems[id];
                const key = `${id}-${index}`;

                return (
                  <MultiSelectListItem
                    className="multiselect-list-item"
                    label={label}
                    key={key}
                    id={id}
                    name={name}
                    handleInputChange={handleInputChange}
                    checked={checked}
                    forwardRef={(el) => {
                      this.listItems[index] = el;
                      return el;
                    }}
                  />
                );
              })}
            </ListboxKeyNav>
            {onSelectionApplied && (
              <MultiSelectFooter applyButtonText={applyButtonText} handleApplyClick={handleApplyClick} />
            )}
          </div>
        )}
      </div>
    );
  }
}

export default MultiSelect;
