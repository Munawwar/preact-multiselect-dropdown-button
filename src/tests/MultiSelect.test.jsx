import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import MultiSelect from '../lib/MultiSelect/MultiSelect';

describe('<MultiSelect />', () => {
  let container;

  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
  });

  afterEach(() => {
    document.body.removeChild(container);
    container = null;
  });

  const multiSelectProps = {
    list: [
      {
        label: 'First option',
        name: 'first-option',
        id: 'first-option-1',
        checked: false
      },
      {
        label: 'Second option',
        name: 'second-option',
        id: 'second-option-2',
        checked: false
      },
      {
        label: 'Third option',
        name: 'third-option',
        id: 'third-option-3',
        checked: false
      }
    ],
    onSelectionApplied: selection => {
      console.log('Selected : ', selection);
    },
    onOptionChanged: optionState => {
      console.log('optionState : ', optionState);
    },
    dropdownButtonText: 'Selected',
    resetButtonText: 'Reset',
    applyButtonText: 'Apply'
  };

  test('should match snapshot - dropdown closed', () => {
    const wrapper = mount(<MultiSelect {...multiSelectProps} />);

    expect(wrapper.getDOMNode()).toMatchSnapshot();
  });

  test('should match snapshot - dropdown opened', () => {
    const wrapper = mount(<MultiSelect {...multiSelectProps} />);
    wrapper.find('button.multiselect-button-dropdown').simulate('click');

    expect(wrapper.getDOMNode()).toMatchSnapshot();
  });

  test('should open dropdown when button is clicked', () => {
    const wrapper = mount(<MultiSelect {...multiSelectProps} />);

    expect(wrapper.find('button.multiselect-button-dropdown')).toBeDefined();
    expect(wrapper.find('.multiselect-section-wrapper').get(0)).toBeUndefined();

    wrapper.find('button.multiselect-button-dropdown').simulate('click');

    expect(wrapper.find('.multiselect-section-wrapper')).toBeDefined();
  });

  test('should select the first item clicking on its input checkbox', () => {
    act(() => {
      ReactDOM.render(<MultiSelect {...multiSelectProps} />, container);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const firstOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[0];
    const firstOptionCheckbox = firstOptionLabel.querySelector('input.multiselect-list-item-checkbox');

    act(() => {
      firstOptionCheckbox.dispatchEvent(new MouseEvent('click'));
    });

    expect(firstOptionCheckbox.checked).toBeTruthy();
  });

  test('should select the first item clicking on its label', () => {
    act(() => {
      ReactDOM.render(<MultiSelect {...multiSelectProps} />, container);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const firstOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[0];
    const firstOptionCheckbox = firstOptionLabel.querySelector('input.multiselect-list-item-checkbox');

    act(() => {
      firstOptionLabel.dispatchEvent(new KeyboardEvent('click'));
    });

    expect(firstOptionCheckbox.checked).toBeTruthy();
  });

  test('should unselect the first item', () => {
    act(() => {
      ReactDOM.render(<MultiSelect {...multiSelectProps} />, container);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const firstOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[0];
    const firstOptionCheckbox = firstOptionLabel.querySelector('input.multiselect-list-item-checkbox');

    act(() => {
      firstOptionLabel.dispatchEvent(new KeyboardEvent('click'));
    });

    expect(firstOptionCheckbox.checked).toBeTruthy();

    act(() => {
      firstOptionLabel.dispatchEvent(new KeyboardEvent('click'));
    });

    expect(firstOptionCheckbox.checked).toBeFalsy();
  });

  test('should apply selections', () => {
    const spyApply = jest.spyOn(multiSelectProps, 'onSelectionApplied');
    act(() => {
      ReactDOM.render(<MultiSelect {...multiSelectProps} />, container);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const firstOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[0];
    act(() => {
      firstOptionLabel.dispatchEvent(new KeyboardEvent('click'));
    });

    const applyButton = container.querySelector('button.multiselect-apply-button');
    act(() => {
      applyButton.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    expect(spyApply).toHaveBeenCalledWith({
      'first-option-1': true,
      'second-option-2': false,
      'third-option-3': false
    });
  });

  test('should badge have the right quantity', () => {
    act(() => {
      ReactDOM.render(<MultiSelect {...multiSelectProps} />, container);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const firstOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[0];
    act(() => {
      firstOptionLabel.dispatchEvent(new KeyboardEvent('click'));
    });

    const badge = container.querySelector('span.multiselect-badge');
    expect(badge.textContent).toBe('1');

    const lastOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[2];
    act(() => {
      lastOptionLabel.dispatchEvent(new KeyboardEvent('click'));
    });

    expect(badge.textContent).toBe('2');
  });

  test('should select all items', () => {
    const spyOptionChanged = jest.spyOn(multiSelectProps, 'onOptionChanged');
    act(() => {
      ReactDOM.render(<MultiSelect {...multiSelectProps} />, container);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const buttonSelectAll = container.querySelector('button.multiselect-button-select-all');
    act(() => {
      buttonSelectAll.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    expect(spyOptionChanged).toHaveBeenCalledWith({
      'first-option-1': true,
      'second-option-2': true,
      'third-option-3': true
    });
  });

  test('should reset selection', () => {
    const spyOptionChanged = jest.spyOn(multiSelectProps, 'onOptionChanged');
    act(() => {
      ReactDOM.render(<MultiSelect {...multiSelectProps} />, container);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const buttonSelectAll = container.querySelector('button.multiselect-button-select-all');
    act(() => {
      buttonSelectAll.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    expect(spyOptionChanged).toHaveBeenCalledWith({
      'first-option-1': true,
      'second-option-2': true,
      'third-option-3': true
    });

    act(() => {
      buttonSelectAll.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    expect(spyOptionChanged).toHaveBeenCalledWith({
      'first-option-1': false,
      'second-option-2': false,
      'third-option-3': false
    });
  });

  test('should trigger option changed', () => {
    const spyOptionChanged = jest.spyOn(multiSelectProps, 'onOptionChanged');
    act(() => {
      ReactDOM.render(<MultiSelect {...multiSelectProps} />, container);
    });

    const buttonDropdown = container.querySelector('button.multiselect-button-dropdown');
    act(() => {
      buttonDropdown.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    const firstOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[0];
    act(() => {
      firstOptionLabel.dispatchEvent(new KeyboardEvent('click'));
    });

    expect(spyOptionChanged).toHaveBeenCalledWith({
      'first-option-1': true,
      'second-option-2': false,
      'third-option-3': false
    });

    const lastOptionLabel = container.querySelectorAll('label.multiselect-list-item-label')[2];
    act(() => {
      lastOptionLabel.dispatchEvent(new KeyboardEvent('click'));
    });

    expect(spyOptionChanged).toHaveBeenCalledWith({
      'first-option-1': true,
      'second-option-2': false,
      'third-option-3': true
    });
  });
});
