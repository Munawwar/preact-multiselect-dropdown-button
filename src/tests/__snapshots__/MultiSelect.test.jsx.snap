// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiSelect /> should match snapshot - dropdown closed 1`] = `
<div
  class="MultiSelect-wrapper"
>
  <button
    aria-haspopup="true"
    class="MultiSelect-dropdownButton"
  >
    Selected
     
    <span
      class="MultiSelect-badge"
    >
      0
    </span>
  </button>
</div>
`;

exports[`<MultiSelect /> should match snapshot - dropdown opened 1`] = `
<div
  class="MultiSelect-wrapper"
>
  <button
    aria-haspopup="true"
    class="MultiSelect-dropdownButton MultiSelect-dropdownButton--opened"
  >
    Selected
     
    <span
      class="MultiSelect-badge"
    >
      0
    </span>
  </button>
  <div
    class="MultiSelect-sectionWrapper"
  >
    <div
      class="MultiSelect-listButtonsWrapper"
    >
      <button
        class="MultiSelect-listButtonsSelectAll"
      >
        Select All
      </button>
      <button
        class="MultiSelect-listButtonsReset"
      >
        Reset
      </button>
    </div>
    <ul
      class="MultiSelect-list MultiSelect-list--hasFooter"
    >
      <li
        class="MultiSelect-listItem"
        role="option"
      >
        <label
          aria-checked="false"
          class="MultiSelect-listItemLabel"
          for="first-option-1"
          role="checkbox"
          tabindex="0"
        >
          First option
          <input
            class="MultiSelect-listItemCheckbox"
            id="first-option-1"
            name="first-option"
            type="checkbox"
          />
        </label>
      </li>
      <li
        class="MultiSelect-listItem"
        role="option"
      >
        <label
          aria-checked="false"
          class="MultiSelect-listItemLabel"
          for="second-option-2"
          role="checkbox"
          tabindex="0"
        >
          Second option
          <input
            class="MultiSelect-listItemCheckbox"
            id="second-option-2"
            name="second-option"
            type="checkbox"
          />
        </label>
      </li>
      <li
        class="MultiSelect-listItem"
        role="option"
      >
        <label
          aria-checked="false"
          class="MultiSelect-listItemLabel"
          for="third-option-3"
          role="checkbox"
          tabindex="0"
        >
          Third option
          <input
            class="MultiSelect-listItemCheckbox"
            id="third-option-3"
            name="third-option"
            type="checkbox"
          />
        </label>
      </li>
    </ul>
    <div
      class="MultiSelect-footer"
    >
      <button
        class="MultiSelect-footerApplyButton"
      >
        Apply
      </button>
    </div>
  </div>
</div>
`;
